contracts/index_executor/src/storage.rs

```diff
@@ -0,0 +1,323 @@
+    pub read_time: String,
+}
+
+/// The client that interacts with cloud storage services.
+/// - `url`: The base URL of the service REST endpoint
+/// - `key`: The credentials associated with the service, generally the access token
+///
+/// By concacting the `url` with the specific API path, we can get a full REST API endpoint.
+/// The result of every request will returned immediately along with the HTTP response.
+///
+/// **Note** The `key` should be kept secret when passing from other modules
+pub struct StorageClient {
+    url: String,
+    key: String,
+}
+
+impl StorageClient {
+    pub fn new(url: String, key: String) -> Self {
+        StorageClient { url, key }
+    }
+
+    /// Send a request to the storage service according to the REST API specification
+    fn send_request(
+        &self,
+        method: &str,
+        api: &str,
+        request: &str,
+    ) -> Result<Vec<u8>, &'static str> {
+        let content_length = format!("{}", request.len());
+        let access_key = format!("Bearer {}", self.key);
+        let headers: Vec<(String, String)> = vec![
+            ("Content-Type".into(), "application/json".into()),
+            ("Authorization".into(), access_key),
+            ("Content-Length".into(), content_length),
+        ];
+
+        let response: pink_extension::chain_extension::HttpResponse = http_req!(
+            method,
+            format!("{}{}", self.url.clone(), api),
+            request.as_bytes().to_vec(),
+            headers
+        );
+        if response.status_code != 200 {
+            return Err("CallServiceFailed");
+        }
+
+        Ok(response.body)
+    }
+
+    /// Return (encoded_data, document_id) if success
+    fn read_storage(&self, key: &[u8]) -> Result<Option<(Vec<u8>, String)>, &'static str> {
+        let key = key
+            .iter()
+            .map(|byte| format!("{byte:02x}"))
+            .collect::<String>();
+        pink_extension::debug!("read_storage: id: {}", key);
+
+        let cmd = format!(
+            r#"{{
+                "structuredQuery": {{
+                    "from": [{{
+                      "collectionId": "index-storage"
+                    }}],
+                    "where": {{
+                      "fieldFilter": {{
+                        "field": {{
+                          "fieldPath": "id"
+                        }},
+                        "op": "EQUAL",
+                        "value": {{
+                          "stringValue": "{key}"
+                        }}
+                      }}
+                    }},
+                    "limit": 1
+                  }}
+            }}"#
+        );
+
+        let response_body: Vec<u8> = self.send_request("POST", "documents:runQuery", &cmd)?;
+        if let Ok(response) = pink_json::from_slice::<Vec<ResponseData>>(&response_body) {
+            Ok(if !response.is_empty() {
+                let data_str = response[0].document.fields.data.string_value.clone();
```

Comment:
Please follow the "return early" style error handling. So instead of write if like this:

```
if ok1 {
  if ok2 {
    return Ok()
  } else {
    return Err()
  }
} else {
  if ok3 {
    return Ok()
  } else {
    return Err()
  }
}
```

You can return the error earlier to make the indent level less:

```
if !ok1 {
  if !ok3 {
    return Err()
  }
  return Ok()
}
if !ok2 {
  return Err()
}
Ok()
```

Reference: https://medium.com/swlh/return-early-pattern-3d18a41bba8

Reply:
Done